File: libraries/ios/Sources/VoidhashCore/Caching/CacheNamespace.swift
import Foundation

/// Builds the storage key prefix isolating one SDK configuration from every other writer.
///
/// `UserDefaults` is shared with the host app, so a bare `voidhash:` prefix can collide with the
/// host's own keys and, worse, two clients configured with different publishable keys or origins
/// would read each other's entries. The namespace is
/// `vh:<schemaVersion>:<hash(publishableKey + baseUrl)>:`; bumping ``schemaVersion`` lets a future
/// release abandon the whole namespace without migration heuristics.
public enum CacheNamespace {
    /// Version of the persisted cache layout.
    public static let schemaVersion = 1

    /// Returns the prefix every key written for this configuration carries.
    public static func prefix(publishableKey: String, baseUrl: URL) -> String {
        let fingerprint = hash(publishableKey + "|" + baseUrl.absoluteString)
        return "vh:\(schemaVersion):\(fingerprint):"
    }

    /// 32-bit FNV-1a over UTF-8, rendered as eight lowercase hex digits.
    ///
    /// Deliberately not `hashValue`: Swift seeds that per process, so it would produce a
    /// different namespace on every launch. The width and rendering match the Kotlin and
    /// TypeScript SDKs so an embedded client and its host SDK derive one namespace.
    public static func hash(_ value: String) -> String {
        var result: UInt32 = 0x811c_9dc5
        for byte in Array(value.utf8) {
            result ^= UInt32(byte)
            result = result &* 0x0100_0193
        }
        let hex = String(result, radix: 16)
        return String(repeating: "0", count: max(0, 8 - hex.count)) + hex
    }
}

File: libraries/android/sdk/src/main/java/com/voidhash/sdk/cache/CacheNamespace.kt
package com.voidhash.sdk.cache

/** Version of the on-disk cache layout; bumping it retires every entry written before it. */
const val CACHE_SCHEMA_VERSION: Int = 1

/**
 * Builds the storage prefix for one project: `vh:<version>:<hash>:`.
 *
 * Two Voidhash clients in the same process (or a key rotation between builds) must not read
 * each other's people, flags and paywalls, so the publishable key and the API origin are
 * folded into the prefix. The version segment lets a later SDK release discard or migrate a
 * whole layout without having to recognise individual keys.
 */
fun cacheKeyPrefix(publishableKey: String, baseUrl: String): String =
    "vh:$CACHE_SCHEMA_VERSION:${fnv1aHex("$publishableKey|$baseUrl")}:"

/**
 * 32-bit FNV-1a, rendered as eight lowercase hex digits.
 *
 * The prefix has to be identical for the same inputs across processes and releases, which
 * rules out `String.hashCode` (unspecified across platforms) and anything seeded.
 */
internal fun fnv1aHex(value: String): String {
    var hash = 0x811C9DC5u
    for (byte in value.toByteArray(Charsets.UTF_8)) {
        hash = hash xor (byte.toUInt() and 0xFFu)
        hash *= 0x01000193u
    }
    return hash.toString(16).padStart(8, '0')
}
